other/upgrade_2-1_mysql.sql

```diff
@@ -2045,157 +2045,208 @@ ADD COLUMN in_inbox TINYINT(3) NOT NULL DEFAULT '1';
+				$labels = explode(',', $row['labels']);
 
-				$new_label_info = $label_info_2[$row['id_member']][$a_label];
-				$inserts[] = array($row['id_pm'], $new_label_info);
-			}
-		}
+				foreach ($labels as $a_label)
+				{
+					if ($a_label == '-1')
+						continue;
 
-		$smcFunc['db_free_result']($get_pm_labels);
+					$new_label_info = $label_info_2[$row['id_member']][$a_label];
+					$inserts[] = array($row['id_pm'], $new_label_info);
+				}
+			}
 
-		// Insert the new data
-		if (!empty($inserts))
-		{
-			$smcFunc['db_insert']('', '{db_prefix}pm_labeled_messages', array('id_pm' => 'int', 'id_label' => 'int'), $inserts, array());
-		}
+			$smcFunc['db_free_result']($get_pm_labels);
 
-		// Final step of this ridiculously massive process
-		$get_pm_rules = $smcFunc['db_query']('', '
-			SELECT id_member, id_rule, actions
-			FROM {db_prefix}pm_rules',
-			array(
-			)
-		);
+			// Insert the new data
+			if (!empty($inserts))
+			{
+				$smcFunc['db_insert']('', '{db_prefix}pm_labeled_messages', array('id_pm' => 'int', 'id_label' => 'int'), $inserts, array());
+			}
 
-		// Go through the rules, unserialize the actions, then figure out if there's anything we can use
-		while ($row = $smcFunc['db_fetch_assoc']($get_pm_rules))
-		{
-			// Turn this INTo an array...
-			$actions = unserialize($row['actions']);
+			// Final step of this ridiculously massive process
+			$get_pm_rules = $smcFunc['db_query']('', '
+				SELECT id_member, id_rule, actions
+				FROM {db_prefix}pm_rules
+				WHERE id_member IN ({array_int:member_list})',
+				array(
+					'member_list' => $member_list,
+				)
+			);
 
-			// Loop through the actions and see if we're applying a label anywhere
-			foreach ($actions as $index => $action)
+			// Go through the rules, unserialize the actions, then figure out if there's anything we can use
+			while ($row = $smcFunc['db_fetch_assoc']($get_pm_rules))
 			{
-				if ($action['t'] == 'lab')
+				$updated = false;
+
+				// Turn this into an array...
+				$actions = unserialize($row['actions']);
+
+				// Loop through the actions and see if we're applying a label anywhere
+				foreach ($actions as $index => $action)
 				{
-					// Update the value of this label...
-					$actions[$index]['v'] = $label_info_2[$row['id_member']][$action['v']];
+					if ($action['t'] == 'lab')
+					{
+						// Update the value of this label...
+						$actions[$index]['v'] = $label_info_2[$row['id_member']][$action['v']];
+						$updated = true;
+					}
+				}
+
+				if ($updated)
+				{
+					// Put this back into a string
+					$actions = serialize($actions);
+
+					$smcFunc['db_query']('', '
+						UPDATE {db_prefix}pm_rules
+						SET actions = {string:actions}
+						WHERE id_rule = {int:id_rule}',
+						array(
+							'actions' => $actions,
+							'id_rule' => $row['id_rule'],
+						)
+					);
 				}
 			}
 
-			// Put this back into a string
-			$actions = serialize($actions);
+			$smcFunc['db_free_result']($get_pm_rules);
 
```

Comment:
Good idea.  Only reason sm website made it through is because we used the CLI to upgrade, avoiding any issues with timeouts.  Although we had some crashes in the upgrader that have now been fixed.

Reply:
Done